source/trace_event.c

```diff
@@ -0,0 +1,191 @@
+
+struct aws_trace_event *trace_event;
+// uint64_t listener_id = 1;
+uint64_t start_time;
+
+/*
+ * Private API
+ */
+
+#define ERROR_CHECK(result)                                                                                            \
+    do {                                                                                                               \
+        if (!result) {                                                                                                 \
+            AWS_FATAL_ASSERT(0 && "trace failure")                                                                     \
+        }                                                                                                              \
+    } while (0)
+
+struct aws_trace_event {
+    struct cJSON *root, *event_array;
+    struct aws_bus bus;
+    struct aws_allocator *allocator;
+};
+
+/* Add trace event meta data to JSON object when the message bus allows it */
+void aws_trace_event_listener(uint64_t address, const void *msg, void *user_data) {
+    struct aws_trace_event_metadata *trace_event_data = (struct aws_trace_event_metadata *)msg;
+    if (trace_event_data == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    cJSON *event = cJSON_CreateObject();
+    if (event == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+
+    cJSON_AddItemToArray(trace_event->event_array, event);
+
+    /* add options for args later on, will initially be empty for now */
+    cJSON_AddItemToObject(event, "args", cJSON_CreateObject());
+
+    /* add more options later on */
+    if (cJSON_AddStringToObject(event, "cat", trace_event_data->category) == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    if (cJSON_AddStringToObject(event, "name", trace_event_data->name) == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    /* Fix for buffer overflow issue using char phase reference */
+    char ph[2];
+    strncat(ph, &(trace_event_data->phase), 1);
+    if (cJSON_AddStringToObject(event, "ph", ph) == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    if (cJSON_AddNumberToObject(event, "pid", trace_event_data->process_id) == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    if (cJSON_AddNumberToObject(event, "tid", trace_event_data->thread_id)) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+    if (cJSON_AddNumberToObject(event, "ts", (double)trace_event_data->timestamp)) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
+}
+
+/* Destructor of trace event metadata sent through the bus */
+void aws_trace_event_destroy(void *payload) {}
+
+void aws_trace_system_write(int time_unit, const char *filename) {
+    if (time_unit) {
+        if (cJSON_AddStringToObject(trace_event->root, "displayTimeUnit", "ns") == NULL) {
+            aws_raise_error(AWS_ERROR_OOM);
+        }
+    }
+    char *out = cJSON_Print((trace_event->root));
+    if (out == NULL) {
+        aws_raise_error(AWS_ERROR_OOM);
+    }
```

Comment:
calling `aws_raise_error()` doesn't really do anything. It sets a thread-local variable, but it doesn't automatically `return` or anything, this function is going to keep running. If this is in error, you'll want to either FATAL_ASSERT, or `return` or `goto` somewhere.

Seems like most of the error-handling in this file needs to be fixed up accordingly